### DISCUSSION TITLE
Synchronous persisted store doesn't load data from localStorage on the first render

### DISCUSSION THREAD

**Discussion by filiptdz** [pmndrs/zustand#2735] [Bug report]
### Bug Description

I'm using NextJS 14.1.0, with Zustand 4.5.5

When I load a persisted store, even though it should be loaded synchronously from localStorage, the first render on the client returns the default values.

This is relevant for a client component on NextJS. It'll work fine on a production build (where the component will in fact only render on the client) but the bug happens while running the NextJS dev server, which server-renders every component regardless of whether it's a server component or not.

Here's a minimal repro: https://codesandbox.io/p/devbox/currying-morning-skw596

If you continuously reload the page, you'll see the data is always loaded with the default value on the first render, which makes it hard to deal with `useEffect`s that depend on the initial value. A loading state could be used, but should be unnecessary, considering the storage is synchronous.

Interestingly, if you uncomment line 22, and call `store.persist.hasHydrated()` on the server, it works as expected when you get to the client.

Could something similar be done as part of the implementation of `persist`?

### Reproduction Link

https://codesandbox.io/p/devbox/currying-morning-skw596

**Comment by dbritto-dev** [comment]
@filiptdz here you go -> https://stackblitz.com/edit/nextjs-and-zustand-persist?file=app%2F_providers%2Fcounter-store-provider.tsx

Few things about your implementation that we talked about in our nextjs guide:
- Try not to use global stores (`create`/`createWithEqualityFn`)
- Try to use scoped stores -> Context API (`Context` + `useContext`) + `createStore`
- Skip hydration on server and hydrate on client `skipHydration: true` and `useEffect`/`useLayoutEffect` + `persist.hydrate()`, I recommended to use `useLayoutEffect` over `useEffect` but you can use `useEffect` too
- Use providers (`Context.Provider`) on your root layout or page layout (if you want to just create the store for just that page)

**Reply by filiptdz** [reply]
Thank you for your response.

I see an issue with the repro I provided - I hadn't realized before, but it just breaks server rendering, which is why this solution "works".

The documentation and example you provided suggest treating localStorage asynchronously, which doesn't really make sense to me. Is there a reason why we can't load state on the first pass, such that a loading state never needs to be displayed?

**Reply by dbritto-dev** [reply]
@filiptdz few things:

- There's no `localStorage` on server
- Server and client should the same at first render
- The `getItem` from `localStorage` should be treated as async operation due to hydration mechanism
- There's no way to avoid the loading state if you want to use persist middleware with any storage

**Reply by dbritto-dev** [reply]
@filiptdz for SPA apps you don't need to worry about that but on SSR apps you have to

**Reply by filiptdz** [reply]
Those are great points. I'm using NextJS SSR for the initial page load, but all the pages are client components, so after the initial page load, it behaves much like a SPA.

Weirdly, when that's built an run with `next start`, I can see the store loads the values from localStorage immediately. If I run `next dev`, during the first render we get the default values, which are then replaced by the localStorage values on the very next render.

So far I can't see any explanation for this, besides `useSyncExternalStore` behaving differently or something.

It looks more like a NextJS issue than Zustand though, so I'll dig deeper and let you know if I can find anything that could benefit this project.

**Reply by dbritto-dev** [reply]
@filiptdz is not an issue that's expected something people forgot that there's not store on server so you couldn't read or write to store and when they built their apps they get those weird behaviors. I encourage you to read our nextjs guide and hydration guide to understand everything

**Comment by filiptdz** [comment]
not something I expect to be changed on the main branch, but I wasn't interested in this behavior where the first render on the client uses an empty state, and have to deal with loading states just for that first render.

Instead, I would rather have the local storage from the first client render, and deal with hydration errors if they happen. To achieve this, I created a patch replacing all the calls to `useSyncExternalStoreWithSelector` with `api.getState` instead of `api.getInitialState`, so something like this:
```
useSyncExternalStoreWithSelector(
     api.subscribe,
     api.getState,
-    api.getServerState || api.getInitialState,
+    api.getState,
     selector,
     equalityFn
   )
```

I also had some issues with `'use-sync-external-store/shim/with-selector.js'`, so I replaced those imports with `'use-sync-external-store/with-selector.js'` since I'm using a modern React version. I couldn't replicate this last issue on a fresh project, so it's likely an issue with the dependencies on my project, but might be useful to anyone else running into the same problems.
